org/mozilla/jss/ssl/javax/JSSEngineReferenceImpl.java

```diff
@@ -0,0 +1,1170 @@
+        }
+
+        debug("JSSEngine.initServer(): " + cert);
+        debug("JSSEngine.initServer(): " + key);
+
+        // Configure SSL server with the given certificate and its private
+        // key.
+        if (SSL.ConfigServerCert(ssl_fd, cert, key) == SSL.SECFailure) {
+            throw new RuntimeException("Unable to initialize server with cert and key: " + errorText(PR.GetError()));
+        }
+        session.setLocalCertificates(new PK11Cert[]{ cert } );
+
+        // Create a small session cache.
+        //
+        // TODO: Make this configurable.
+        if (SSL.ConfigServerSessionIDCache(1, 100, 100, null) == SSL.SECFailure) {
+            throw new RuntimeException("Unable to configure server session cache: " + errorText(PR.GetError()));
+        }
+
+        // Update handshake status; client initiates connection, so wait
+        // for unwrap on the server end.
+        handshake_state = SSLEngineResult.HandshakeStatus.NEED_UNWRAP;
+
+        // Only specify these on the server side as they affect what we
+        // want from the remote peer in NSS. In the server case, this is
+        // client auth, but if we were to set these on the client, it would
+        // affect server auth.
+        if (SSL.OptionSet(ssl_fd, SSL.REQUEST_CERTIFICATE, want_client_auth || need_client_auth ? 1 : 0) == SSL.SECFailure) {
+            throw new RuntimeException("Unable to configure SSL_REQUEST_CERTIFICATE option: " + errorText(PR.GetError()));
+        }
+
+        if (SSL.OptionSet(ssl_fd, SSL.REQUIRE_CERTIFICATE, need_client_auth ? 1 : 0) == SSL.SECFailure) {
+            throw new RuntimeException("Unable to configure SSL_REQUIRE_CERTIFICATE option: " + errorText(PR.GetError()));
+        }
+
+        step_handshake = true;
+    }
+
+    private void applyCiphers() {
+        debug("JSSEngine: applyCiphers()");
+        // Enabled the ciphersuites specified by setEnabledCipherSuites(...).
+        // When this isn't called, enabled_ciphers will be null, so we'll just
+        // use whatever is enabled by default.
+        if (enabled_ciphers == null) {
+            return;
+        }
+
+        // We need to disable the suite if it isn't present in the list of
+        // suites above. Be lazy about it for the time being and disable all
+        // cipher suites first.
+        for (SSLCipher suite : SSLCipher.values()) {
+            if (SSL.CipherPrefSet(ssl_fd, suite.getID(), false) == SSL.SECFailure) {
+                // warn("Unable to set cipher suite preference for " + suite.name() + ": " + errorText(PR.GetError()));
+            }
+        }
+
+        // Only enable these particular suites. When a cipher suite can't be
+        // enabled it is most likely due to local policy. Log it. Also log
+        // which ciphers were successfully enabled for debugging purposes.
+        for (SSLCipher suite : enabled_ciphers) {
+            if (suite == null) {
+                continue;
+            }
+
```

Comment:
It depends on why we can't enable said cipher. If it is because it isn't allowed by policy, yeah. But if it is because the NSS version is too old? Maybe not. However, NSS doesn't really "blow up" here...

It fails:
 1. When `ssl_fd` is bad (which isn't likely -- we would've blown up a lot earlier unless we're seeing memory corruption)
 2. If the suite doesn't exist (that's discussed above -- and it isn't in the case where the suite was explicitly removed from NSS, like some of the old export grade ones) 

That's it. We handle 2 because we're setting these from `SSLCiphers` and we've previously converted them from String names to `SSLCipher` instances.

This also means that the policy check really needs to occur at the `SSLEngine` level, not at the NSS level. I'll make sure we check that in a separate PR. 

---

Code:

```c
SECStatus
SSL_CipherPrefSet(PRFileDesc *fd, PRInt32 which, PRBool enabled)
{
    sslSocket *ss = ssl_FindSocket(fd);

    if (!ss) {
        SSL_DBG(("%d: SSL[%d]: bad socket in CipherPrefSet", SSL_GETPID(), fd));
        return SECFailure;
    }
    if (ssl_IsRemovedCipherSuite(which))
        return SECSuccess;
    return ssl3_CipherPrefSet(ss, (ssl3CipherSuite)which, enabled);
}

SECStatus
ssl3_CipherPrefSet(sslSocket *ss, ssl3CipherSuite which, PRBool enabled)
{
    ssl3CipherSuiteCfg *suite;

    suite = ssl_LookupCipherSuiteCfgMutable(which, ss->cipherSuites);
    if (suite == NULL) {
        return SECFailure; /* err code was set by ssl_LookupCipherSuiteCfg */
    }
    suite->enabled = enabled;
    return SECSuccess;
}
```